File: benchmarks/cross-language/erlang/counting.erl
-module(counting).
-export([start/0]).

get_messages() ->
    case os:getenv("BENCHMARK_MESSAGES") of
        false -> 100000;
        Val -> list_to_integer(Val)
    end.

start() ->
    Messages = get_messages(),
    io:format("=== Erlang Counting Actor Benchmark ===~n"),
    io:format("Messages: ~p~n~n", [Messages]),

    Parent = self(),
    Counter = spawn(fun() -> counter(0, Parent) end),

    Start = erlang:monotonic_time(nanosecond),

    % Send all messages
    send_messages(Counter, Messages),

    % Signal done and wait for count
    Counter ! {done, self()},
    receive
        {count, Count} ->
            End = erlang:monotonic_time(nanosecond),
            ElapsedNs = End - Start,
            ElapsedSec = ElapsedNs / 1000000000.0,

            if
                Count =/= Messages ->
                    io:format("VALIDATION FAILED: expected ~p, got ~p~n", [Messages, Count]);
                true -> ok
            end,

            NsPerMsg = ElapsedNs / Messages,
            MsgPerSec = Messages / ElapsedSec,
            io:format("ns/msg:         ~.2f~n", [NsPerMsg]),
            io:format("Throughput:     ~.2f M msg/sec~n", [MsgPerSec / 1000000])
    after 60000 ->
        io:format("Timeout!~n"),
        halt(1)
    end,

    halt(0).

send_messages(_Counter, 0) -> ok;
send_messages(Counter, N) ->
    Counter ! increment,
    send_messages(Counter, N - 1).

counter(Count, Parent) ->
    receive
        increment ->
            counter(Count + 1, Parent);
        {done, From} ->
            From ! {count, Count}
    end.

File: benchmarks/cross-language/elixir/counting.exs
# Elixir Counting Actor Benchmark (Savina-style)

defmodule Counting do
  @messages (System.get_env("BENCHMARK_MESSAGES") |> case do
    nil -> 100_000
    val -> String.to_integer(val)
  end)

  def start do
    IO.puts("=== Elixir Counting Actor Benchmark ===")
    IO.puts("Messages: #{@messages}\n")

    parent = self()
    counter = spawn(fn -> counter_loop(0, parent) end)

    start_time = :erlang.monotonic_time(:nanosecond)

    # Send all messages
    Enum.each(1..@messages, fn _ -> send(counter, :increment) end)

    # Signal done
    send(counter, {:done, self()})

    # Wait for count
    receive do
      {:count, count} ->
        end_time = :erlang.monotonic_time(:nanosecond)
        elapsed_ns = end_time - start_time
        elapsed_sec = elapsed_ns / 1_000_000_000.0

        if count != @messages do
          IO.puts("VALIDATION FAILED: expected #{@messages}, got #{count}")
        end

        throughput = @messages / elapsed_sec / 1_000_000
        ns_per_msg = elapsed_ns / @messages

        IO.puts("ns/msg:         #{:io_lib.format("~.2f", [ns_per_msg])}")
        IO.puts("Throughput:     #{:io_lib.format("~.2f", [throughput])} M msg/sec")
    after
      60_000 ->
        IO.puts("Timeout!")
        System.halt(1)
    end
  end

  defp counter_loop(count, parent) do
    receive do
      :increment ->
        counter_loop(count + 1, parent)
      {:done, from} ->
        send(from, {:count, count})
    end
  end
end

Counting.start()
